server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java

```diff
@@ -862,17 +862,14 @@ public boolean moveShards() {
             for (var storedShardMovement : bestNonPreferredShardMovementsTracker.getBestShardMovements()) {
                 final var shardRouting = storedShardMovement.shardRouting();
                 final var index = projectIndex(shardRouting);
-                // If `shardMoved` is true, there may have been moves that have made our previous move decision
-                // invalid, so we must call `decideMove` again. If not, we know we haven't made any moves, and we
-                // can use the cached decision.
-                final var moveDecision = shardMoved ? decideMove(index, shardRouting) : storedShardMovement.moveDecision();
+                final var moveDecision = reassessDecisionIfRequired(index, storedShardMovement, shardMoved);
```

Comment:
opt nit.

This method name doesn't use any terminology that's familiar in this context. Pretty hard to name, really. Perhaps something like recalculateMoveIfNeeded might be better.